### PR TITLE
feat: Arrow/Parquet IO for ProteinIdentification, FeatureMap, and ConsensusMap

### PR DESCRIPTION
**Pull request by timosachsenberg**
## Summary

Adds full round-trip Arrow/Parquet IO for **ProteinIdentification**, **FeatureMap**, and **ConsensusMap**. Each map type exports to a `.cmd` directory containing multiple Parquet files.

### New classes
- **ProteinIdentificationArrowIO** — export/import protein hits, groups, and search parameters
- **FeatureMapArrowIO** — export/import features with convex hulls, nested subordinates, bounding boxes, and PSM linkage
- **ConsensusMapArrowIO** — export/import consensus features with feature handles, column headers, and PSM linkage

### Key design decisions
- Reuses **QPXFile** for PSM export/import (shared across all map types)
- Reuses **ProteinIdentificationArrowIO** for protein data (shared across FeatureMap and ConsensusMap)
- Normalized schema: run-level metadata lives in \`search_params.parquet\`, linked by \`run_identifier\`
- MetaValue list types (\`INT_LIST\`, \`DOUBLE_LIST\`, \`STRING_LIST\`) fully round-tripped
- Per-PSM \`higher_score_better\` column for correct score orientation independent of run-level setting
- Hide Arrow bundled dependency symbols on Linux (\`--exclude-libs\`) to prevent conflicts with pyarrow

### Directory layout

\`\`\`
experiment.cmd/
├── consensus_features.parquet   (or features.parquet)
├── psms.parquet
├── proteins.parquet
├── protein_groups.parquet
└── search_params.parquet
\`\`\`

### Files changed (21 files, +9426 lines)

| Category | Files |
|----------|-------|
| Headers | \`ConsensusMapArrowIO.h\`, \`FeatureMapArrowIO.h\`, \`ProteinIdentificationArrowIO.h\` |
| Sources | \`ConsensusMapArrowIO.cpp\`, \`FeatureMapArrowIO.cpp\`, \`ProteinIdentificationArrowIO.cpp\`, \`QPXFile.cpp\` |
| pyOpenMS bindings | \`ConsensusMapArrowIO.pxd\`, \`FeatureMapArrowIO.pxd\`, \`ProteinIdentificationArrowIO.pxd\` |
| Build config | \`sources.cmake\` (x2), \`CMakeLists.txt\` (x2), \`executables.cmake\`, \`create_cpp_extension.py\` |
| Tests | \`ConsensusMapArrowIO_test.cpp\`, \`FeatureMapArrowIO_test.cpp\`, \`ProteinIdentificationArrowIO_test.cpp\`, \`QPXFile_test.cpp\` |

## Test plan

- [x] \`QPXFile_test\` — PSM export with higher_score_better, scan, reference_file_name
- [x] \`ProteinIdentificationArrowIO_test\` — proteins, groups, search params round-trip
- [x] \`FeatureMapArrowIO_test\` — features, convex hulls, subordinates, PSMs, metavalues
- [x] \`ConsensusMapArrowIO_test\` — consensus features, handles, column headers, PSMs, metavalues
- [x] CI green on all 5 build platforms (Linux g++/clang++/ARM, macOS, Windows)
- [x] CI green on all 20 wheel test combinations (Python 3.10-3.14 x 4 OS)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/openms/include/OpenMS/FORMAT/ConsensusMapArrowIO.h [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+    and ConsensusMap-level metadata.
+
+    @param[in] directory Input directory path containing Parquet files
+    @param[out] cmap ConsensusMap to populate
+    @return true on success, false on error
+  */
+  static bool importFromParquet(
+    const String& directory,
+    ConsensusMap& cmap);
+};
```
[the remark this answers is not in the document]
Comment (timosachsenberg) [reply src/openms/include/OpenMS/FORMAT/ConsensusMapArrowIO.h:137]: Resolved — `src/pyOpenMS/pxds/ConsensusMapArrowIO.pxd` exists and was committed in 75fd7aa421.

File: src/openms/source/FORMAT/FeatureMapArrowIO.cpp [hunk 2]
```diff
@@ -0,0 +1,1648 @@
...
+    auto write_status = parquet::arrow::WriteTable(
+      *table, arrow::default_memory_pool(), outfile,
+      config.row_group_size, writer_props, arrow_props);
+
+    if (!write_status.ok())
+    {
+      OPENMS_LOG_ERROR << "FeatureMapArrowIO: Failed to write Parquet: "
+                       << write_status.ToString() << std::endl;
+      return false;
+    }
```
[the remark this answers is not in the document]
Comment (timosachsenberg) [reply src/openms/source/FORMAT/FeatureMapArrowIO.cpp:564]: Not actionable — `std::endl` with `OPENMS_LOG_*` macros is the consistent convention across the entire OpenMS codebase (388+ occurrences in FORMAT files alone).

File: src/openms/source/FORMAT/ProteinIdentificationArrowIO.cpp [hunk 3]
```diff
@@ -1580,39 +1471,11 @@ bool ProteinIdentificationArrowIO::importProteinsFromArrow(
     }
     else
     {
-      // Create new ProteinIdentification from run-level data in this row
+      // Create minimal ProteinIdentification with just run_identifier
+      OPENMS_LOG_WARN << "ProteinIdentificationArrowIO: run_identifier '" << run_id
+                      << "' not found in search_params; creating minimal ProteinIdentification" << std::endl;
```
[the remark this answers is not in the document]
Comment (timosachsenberg) [reply src/openms/source/FORMAT/ProteinIdentificationArrowIO.cpp:1524]: Not actionable — this is a standard warning log for a normal code path (missing run identifier creates a minimal ProteinIdentification). No issue here.

File: src/openms/source/FORMAT/ConsensusMapArrowIO.cpp [hunk 4]
```diff
@@ -0,0 +1,1574 @@
...
+      (void)handle_rt_b->Append(handle.getRT());
+      (void)handle_mz_b->Append(handle.getMZ());
+      (void)handle_intensity_b->Append(handle.getIntensity());
+      (void)handle_charge_b->Append(static_cast<int32_t>(handle.getCharge()));
+      (void)handle_width_b->Append(handle.getWidth());
+    }
+
+    // metavalues
+    (void)metavalues_builder.Append();
+    appendMetaValues_(cf, mv_name_b, mv_value_b, mv_type_b, mv_struct_b, excluded_mvs);
```
[the remark this answers is not in the document]
Comment (timosachsenberg) [reply src/openms/source/FORMAT/ConsensusMapArrowIO.cpp:1106]: Not actionable — this is a consistent pattern across both FeatureMapArrowIO and ConsensusMapArrowIO. Neither file calls `Reserve()` on nested ListBuilders (handles, metavalues, convex_hulls). Arrow ListBuilders grow dynamically by design. The `(void)` casts are intentional — errors are caught at `Finish()` via `FINISH_OR_RETURN` macros.

File: src/openms/source/FORMAT/ProteinIdentificationArrowIO.cpp [hunk 5]
```diff
@@ -0,0 +1,1902 @@
...
+    bytes[6] = (bytes[6] & 0x0F) | 0x40; // version 4
+    bytes[8] = (bytes[8] & 0x3F) | 0x80; // variant 1
+    char buf[37];
+    std::snprintf(buf, sizeof(buf),
+      "%02x%02x%02x%02x-%02x%02x-%02x%02x-%02x%02x-%02x%02x%02x%02x%02x%02x",
+      bytes[0], bytes[1], bytes[2], bytes[3],
+      bytes[4], bytes[5], bytes[6], bytes[7],
+      bytes[8], bytes[9], bytes[10], bytes[11],
+      bytes[12], bytes[13], bytes[14], bytes[15]);
+    std::string uuid_str(buf);
```
[the remark this answers is not in the document]
Comment (timosachsenberg) [reply src/openms/source/FORMAT/ProteinIdentificationArrowIO.cpp:114]: Acknowledged — the UUID generation is duplicated across the ArrowIO files. This is a refactoring opportunity but not a bug. Each file is self-contained by design. Can be extracted to a shared utility in a follow-up PR.

File: src/openms/source/FORMAT/ConsensusMapArrowIO.cpp [hunk 6]
```diff
@@ -0,0 +1,1682 @@
...
+                  try
+                  {
+                    String s(mv_value);
+                    if (s.hasPrefix("[") && s.hasSuffix("]")) { s = s.substr(1, s.size() - 2); }
+                    auto sl = ListUtils::create<String>(s);
+                    for (auto& e : sl) { e = e.trim(); }
+                    dp.setMetaValue(mv_name, DataValue(sl));
+                  }
+                  catch (...) { dp.setMetaValue(mv_name, DataValue(mv_value)); }
+                }
```
[the remark this answers is not in the document]
Comment (timosachsenberg) [reply src/openms/source/FORMAT/ConsensusMapArrowIO.cpp:401]: Investigated — `ListUtils::create<T>("")` on an empty string returns an empty vector, so empty lists should round-trip correctly. This is an edge case without explicit test coverage but not a bug.

File: src/openms/source/FORMAT/FeatureMapArrowIO.cpp [hunk 7]
```diff
@@ -367,11 +373,43 @@ namespace // anonymous
...
+                  try
+                  {
+                    String s(mv_value);
+                    if (s.hasPrefix("[") && s.hasSuffix("]")) { s = s.substr(1, s.size() - 2); }
+                    auto sl = ListUtils::create<String>(s);
+                    for (auto& e : sl) { e = e.trim(); }
+                    dp.setMetaValue(mv_name, DataValue(sl));
+                  }
+                  catch (...) { dp.setMetaValue(mv_name, DataValue(mv_value)); }
+                }
```
[the remark this answers is not in the document]
Comment (timosachsenberg) [reply src/openms/source/FORMAT/FeatureMapArrowIO.cpp:412]: Same as ConsensusMapArrowIO — empty lists round-trip correctly via `ListUtils::create<T>("")`. Not a bug.

File: src/openms/source/FORMAT/ConsensusMapArrowIO.cpp [hunk 8]
```diff
@@ -0,0 +1,1682 @@
...
+        if (key == "map_index" || key == "size" || key == "unique_id")
+        {
+          std::string num_str;
+          while (pos < json.size() && (std::isdigit(json[pos]) || json[pos] == '-'))
+          {
+            num_str += json[pos++];
+          }
+          if (key == "map_index") map_index = std::stoull(num_str);
+          else if (key == "size") header.size = std::stoull(num_str);
+          else if (key == "unique_id") header.unique_id = std::stoull(num_str);
```
[the remark this answers is not in the document]
Comment (timosachsenberg) [reply src/openms/source/FORMAT/ConsensusMapArrowIO.cpp:626]: Acknowledged — `std::stoull` can throw on malformed input. However, `deserializeColumnHeaders_` only parses JSON that was serialized by `serializeColumnHeaders_` in the same codebase, so malformed input requires a corrupted Parquet file. The exception would propagate to the caller with a clear error. Low-priority hardening for a follow-up.

File: src/openms/source/FORMAT/ConsensusMapArrowIO.cpp [hunk 9]
```diff
@@ -0,0 +1,1574 @@
...
+
+      int64_t start = list_arr->value_offset(row);
+      int64_t end = start + list_arr->value_length(row);
+      for (int64_t k = start; k < end; ++k)
+      {
+        String name = names_arr->GetString(k);
+        double value = values_arr->Value(k);
+        hit.setMetaValue(name, value);
+      }
+    }
```
[the remark this answers is not in the document]
Comment (timosachsenberg) [reply src/openms/source/FORMAT/ConsensusMapArrowIO.cpp:1576]: Acknowledged — `GetFieldByName` can return nullptr in theory. In practice, the struct schema is created by the same export code, so the fields are guaranteed to exist for our own Parquet files. The outer null-check on the column (`col_additional_scores && \!isNull_`) ensures we only reach this code when the struct column is present with valid data. Same pattern in FeatureMapArrowIO. Low-priority hardening for a follow-up.
